skimage/restoration/_nl_means_denoising.pyx

```diff
@@ -0,0 +1,596 @@
+
+cdef eps = 1.e-8
+
+
+@cython.boundscheck(False)
+cdef inline float patch_distance_2d(DTYPE_t [:, :] p1,
+                                    DTYPE_t [:, :] p2,
+                                    DTYPE_t [:, ::] w, int s):
+    cdef int i, j
+    cdef int center = s / 2
+    # Check if central pixel is too different in the 2 patches
+    cdef float tmp_diff = p1[center, center] - p2[center, center]
+    cdef float init = w[center, center] * tmp_diff * tmp_diff
+    if init > 1:
+        return eps
+    cdef float distance = 0
+    for i in range(s):
+        # exp of large negative numbers will be 0, so we'd better stop
+        if distance > 5:
+            return eps
+        for j in range(s):
+            tmp_diff = p1[i, j] - p2[i, j]
+            distance += w[i, j] * tmp_diff * tmp_diff
+    distance = exp(- distance)
+    return distance
+
+
+@cython.boundscheck(False)
+cdef inline float patch_distance_2drgb(DTYPE_t [:, :, :] p1,
+                                       DTYPE_t [:, :, :] p2,
+                                       DTYPE_t [:, ::] w, int s):
+    cdef int i, j
+    cdef int center = s / 2
+    cdef int color
+    cdef float tmp_diff = 0
+    cdef float distance = 0
+    for i in range(s):
+        # exp of large negative numbers will be 0, so we'd better stop
+        if distance > 5:
+            return eps
+        for j in range(s):
+            for color in range(3):
+                tmp_diff = p1[i, j, color] - p2[i, j, color]
+                distance += w[i, j] * tmp_diff * tmp_diff
+    distance = exp(- distance)
+    return distance
+
+
+@cython.boundscheck(False)
+cdef inline float patch_distance_3d(DTYPE_t [:, :, :] p1,
+                                    DTYPE_t [:, :, :] p2,
+                                    DTYPE_t [:, :, ::] w, int s):
+    cdef int i, j, k
+    cdef float distance = 0
+    cdef float tmp_diff
+    for i in range(s):
+        # exp of large negative numbers will be 0, so we'd better stop
+        if distance > 5:
+            return eps
+        for j in range(s):
+            for k in range(s):
+                tmp_diff = p1[i, j, k] - p2[i, j, k]
+                distance += w[i, j, k] * tmp_diff * tmp_diff
+    distance = exp(- distance)
+    return distance
+
+
+@cython.cdivision(True)
+@cython.boundscheck(False)
+def _nl_means_denoising_2d(image, int s=7, int d=13, float h=0.1):
+    """
+    Perform non-local means denoising on 2-D array
+
+    Parameters
+    ----------
+    image: ndarray
+        input data to be denoised
+
+    s: int, optional
+        size of patches used for denoising
+
+    d: int, optional
+        maximal distance in pixels where to search patches used for denoising
+
+    h: float, optional
+        cut-off distance (in gray levels). The higher h, the more permissive
+        one is in accepting patches.
+    """
+    if s % 2 == 0:
+        s += 1  # odd value for symmetric patch
+    cdef int n_x, n_y
+    n_x, n_y = image.shape
```

Comment:
no worries :-). When I recently read your PR #1280 about coordinates it occurred to me that I should change my naming conventions, but I was too lazy to do it. That's one of the beauties of code review that laziness becomes less acceptable :-)

One thing, though, is that I'm not completely convinced by the name "frame" in 3D. How about "slice"? I don't know if you had the time to read my recent comments in #1280.